compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -3170,7 +3190,16 @@ impl std::hash::Hash for Type {
                 alias.hash(state);
                 args.hash(state);
             }
-            Type::TypeVariable(var) | Type::NamedGeneric(var, ..) => var.hash(state),
+            Type::NamedGeneric(NamedGeneric { type_var, implicit: true, .. }) => {
+                // An implicitly added unbound named generic's hash must be the same as any other
+                // implicitly added unbound named generic's hash.
+                if !type_var.borrow().is_unbound() {
+                    type_var.hash(state);
+                }
```

Comment:
Does it though?
We issue implicit NamedGenerics for associated types so this seems like it'd consider the generics from two separate associated types as equal.

Reply:
Good point.

Originally I fixed this issue in a different way: when comparing a method where clause in a impl against its trait, I removed all implicit generics from the comparison (there's an equality comparison somewhere there). But then I thought I could also fix it by changing the equality... maybe I should have done the former?